src/kernel/arch/x86/keyboard.zig

```diff
@@ -0,0 +1,140 @@
+    if (special_sequence or on_print_screen) {
+        // Most special sequences are followed by an extra key release byte
+        expected_releases = 1;
+        switch (key_code) {
+            72 => key_pos = KeyPosition.UP_ARROW,
+            75 => key_pos = KeyPosition.LEFT_ARROW,
+            77 => key_pos = KeyPosition.RIGHT_ARROW,
+            80 => key_pos = KeyPosition.DOWN_ARROW,
+            // First byte sent for the pause key
+            29 => return @ptrToInt(ctx),
+            42 => {
+                // The print screen key is followed by five extra key release bytes
+                key_pos = KeyPosition.PRINT_SCREEN;
+                on_print_screen = true;
+                expected_releases = 5;
+            },
+            // Second and final byte sent for the pause key
+            69 => {
+                // The pause key is followed by two extra key release bytes
+                key_pos = KeyPosition.PAUSE;
+                expected_releases = 2;
+            },
+            82 => key_pos = KeyPosition.INSERT,
+            71 => key_pos = KeyPosition.HOME,
+            73 => key_pos = KeyPosition.PAGE_UP,
+            83 => key_pos = KeyPosition.DELETE,
+            79 => key_pos = KeyPosition.END,
+            81 => key_pos = KeyPosition.PAGE_DOWN,
+            53 => key_pos = KeyPosition.KEYPAD_SLASH,
+            28 => key_pos = KeyPosition.KEYPAD_ENTER,
+            56 => key_pos = KeyPosition.RIGHT_ALT,
+            91 => key_pos = KeyPosition.SPECIAL,
+            else => return @ptrToInt(ctx),
+        }
+    }
+    key_pos = key_pos orelse switch (key_code) {
+        1 => KeyPosition.ESC,
+        // Number keys and second row
+        2...28 => @intToEnum(KeyPosition, @enumToInt(KeyPosition.ONE) + (key_code - 2)),
+        29 => KeyPosition.LEFT_CTRL,
+        30...40 => @intToEnum(KeyPosition, @enumToInt(KeyPosition.A) + (key_code - 30)),
+        41 => KeyPosition.BACKTICK,
+        42 => KeyPosition.LEFT_SHIFT,
+        43 => KeyPosition.HASH,
+        44...54 => @intToEnum(KeyPosition, @enumToInt(KeyPosition.Z) + (key_code - 44)),
+        55 => KeyPosition.KEYPAD_ASTERISK,
+        56 => KeyPosition.LEFT_ALT,
+        57 => KeyPosition.SPACE,
+        58 => KeyPosition.CAPS_LOCK,
+        69 => KeyPosition.NUM_LOCK,
+        70 => KeyPosition.SCROLL_LOCK,
+        71...73 => @intToEnum(KeyPosition, @enumToInt(KeyPosition.KEYPAD_7) + (key_code - 71)),
+        74 => KeyPosition.KEYPAD_MINUS,
+        75...77 => @intToEnum(KeyPosition, @enumToInt(KeyPosition.KEYPAD_4) + (key_code - 75)),
+        78 => KeyPosition.KEYPAD_PLUS,
+        79...81 => @intToEnum(KeyPosition, @enumToInt(KeyPosition.KEYPAD_1) + (key_code - 79)),
+        82 => KeyPosition.KEYPAD_0,
+        83 => KeyPosition.KEYPAD_DOT,
+        86 => KeyPosition.BACKSLASH,
+        else => null,
+    };
+    if (key_pos) |k| {
+        pressed_keys += 1;
+        _ = keyboard.writeKey(.{ .position = k, .released = released });
```

Comment:
Could be worth logging to indicate if there is no room for the key. Maybe a `notice`?